include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +144,945 @@ struct Basic
 namespace surface_interactions
 {
 
-template<class RayDirInfo>
-struct Isotropic
+#define NBL_CONCEPT_NAME Isotropic
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (iso, T)
+#define NBL_CONCEPT_PARAM_1 (normV, typename T::ray_dir_info_type)
+#define NBL_CONCEPT_PARAM_2 (normN, typename T::vector3_type)
+NBL_CONCEPT_BEGIN(3)
+#define iso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+#define normV NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_1
+#define normN NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_2
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_TYPE)(T::ray_dir_info_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::vector3_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((iso.V), ::nbl::hlsl::is_same_v, typename T::ray_dir_info_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((iso.N), ::nbl::hlsl::is_same_v, typename T::vector3_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((iso.NdotV), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((iso.NdotV2), ::nbl::hlsl::is_same_v, typename T::scalar_type))
```

Comment:
I'm considering one thing, maybe in the concepts we should check for "getters", i.e. functions that get particular things, not variables

Because then the actual implementation has a choice whether to precompute and hold in a member OR to compute on demand, which becomes important later on when; implementing schedulers, dumping state to buffers, or trying to fight register pressure with extra ALU.

Reply:
this requires adding the getters to `SIsotropic` and using them instead of the members ( testing purposes in C++ you can temporarily make them private)